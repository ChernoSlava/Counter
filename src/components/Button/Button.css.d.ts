export const Button: string;
export const Button_increase: string;
export const Button_decrease: string;
export const Button_reset: string;
