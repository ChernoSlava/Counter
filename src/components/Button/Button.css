.Button {
    border: 1px solid aqua;
    border-radius: 6px;
    background-color: transparent;
    width: 140px;
    height: 47px;
    cursor: pointer;
}

.Button_increase {
background-color: aquamarine;
}

.Button_decrease {
    background-color: indianred;
}

.Button_reset {
    background-color: darkgrey;
}