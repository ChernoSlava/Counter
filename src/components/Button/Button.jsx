import React from "react";
import styles from './Button.css';

const typeCss = {
    'increase': styles.Button_increase,
    'decrease': styles.Button_decrease,
    'reset': styles.Button_reset
};

export const Button = ({ title, type }) => {
    return (
        <button className={`${styles.Button} ${typeCss[type] || styles.Button_increase}`}>{title}</button>
    )
}