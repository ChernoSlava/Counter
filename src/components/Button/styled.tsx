import styled from 'styled-components';
import { ThemeType } from '@theme';

export type ButtonType = 'increase' | 'decrease' | 'reset';

export const ButtonStyled = styled.button<{
  renderType: ButtonType;
}>`
  border: 1px solid aqua;
  border-radius: 6px;
  background-color: transparent;
  width: 140px;
  height: 47px;
  cursor: pointer;
  ${props =>
    props.renderType === 'increase' &&
    `
        background-color: ${(props.theme as ThemeType).colors.aqua10};
    `}
  ${props =>
    props.renderType === 'decrease' &&
    `
        background-color: ${(props.theme as ThemeType).colors.red10};
    `}
    ${props =>
    props.renderType === 'reset' &&
    `
        background-color: ${(props.theme as ThemeType).colors.grey10};
    `}
    &:hover {
    opacity: 0.5;
  }
`;
