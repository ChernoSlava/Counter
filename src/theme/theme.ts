export const theme = {
  colors: {
    grey10: 'darkgrey',
    aqua10: 'aquamarine',
    red10: 'indianred',
    red20: 'red',
    black: 'black',
    green10: 'green',
  },
};

export type ThemeType = typeof theme;
