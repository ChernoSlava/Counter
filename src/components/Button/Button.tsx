import React from 'react';
import { ButtonStyled, ButtonType } from './styled';

export type ButtonProps = {
  title?: string;
  type: ButtonType;
  onClick?: () => void;
  isDisabled?: boolean;
};
export const Button: React.FC<ButtonProps> = ({
  title,
  type,
  isDisabled,
  onClick,
}): JSX.Element => {
  return (
    <ButtonStyled disabled={isDisabled} renderType={type} onClick={onClick}>
      {title}
    </ButtonStyled>
  );
};
