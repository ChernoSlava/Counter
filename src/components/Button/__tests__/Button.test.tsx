/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable no-undef */
import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { ThemeProvider } from 'styled-components';

import { theme } from '../../../theme';
import { Button, ButtonProps } from '../Button';

const getUI = (props?: ButtonProps): JSX.Element => {
  return (
    <ThemeProvider theme={theme}>
      <Button
        title="I am button"
        type="decrease"
        onClick={jest.fn()}
        {...(props || {})}
      />
    </ThemeProvider>
  );
};

describe('Button tests', () => {
  it('should render correct decrease button', () => {
    const { container } = render(getUI());

    expect(container).toMatchSnapshot();
  });
  it('should render correct increase button', () => {
    const { container } = render(
      getUI({
        type: 'increase',
      }),
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correct reset button', () => {
    const { container } = render(
      getUI({
        type: 'reset',
      }),
    );

    expect(container).toMatchSnapshot();
  });

  it('should handle onClick', () => {
    const onClick = jest.fn();

    const { getByText } = render(
      getUI({
        type: 'reset',
        onClick,
      }),
    );

    const button = getByText('I am button');
    fireEvent.click(button);

    expect(onClick).toBeCalled();
    expect(onClick).toBeCalledTimes(1);
  });

  it('should not handle onClick', () => {
    const onClick = jest.fn();

    const { getByText } = render(
      getUI({
        type: 'reset',
        onClick,
        isDisabled: true,
      }),
    );

    const button = getByText('I am button');
    fireEvent.click(button);

    expect(onClick).not.toBeCalled();
    expect(onClick).toBeCalledTimes(0);
  });
});
