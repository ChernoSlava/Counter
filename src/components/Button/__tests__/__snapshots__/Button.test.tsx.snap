// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button tests should render correct decrease button 1`] = `
.c0 {
  border: 1px solid aqua;
  border-radius: 6px;
  background-color: transparent;
  width: 140px;
  height: 47px;
  cursor: pointer;
  background-color: indianred;
}

.c0:hover {
  opacity: 0.5;
}

<div>
  <button
    class="c0"
  >
    I am button
  </button>
</div>
`;

exports[`Button tests should render correct increase button 1`] = `
.c0 {
  border: 1px solid aqua;
  border-radius: 6px;
  background-color: transparent;
  width: 140px;
  height: 47px;
  cursor: pointer;
  background-color: aquamarine;
}

.c0:hover {
  opacity: 0.5;
}

<div>
  <button
    class="c0"
  >
    I am button
  </button>
</div>
`;

exports[`Button tests should render correct reset button 1`] = `
.c0 {
  border: 1px solid aqua;
  border-radius: 6px;
  background-color: transparent;
  width: 140px;
  height: 47px;
  cursor: pointer;
  background-color: darkgrey;
}

.c0:hover {
  opacity: 0.5;
}

<div>
  <button
    class="c0"
  >
    I am button
  </button>
</div>
`;
